Store sharding function used in the repo.

kevina opened the pull request:
Store the sharding function used in the file "SHARDING" in the repo.
To make this work the sharding function is now always specified as a
string.

whyrusleeping commented:
What are some example strings? We should make sure that these are somewhat nice multicodecs, cc @Kubuxu

kevina commented on flatfs.go:
```diff
@@ -38,37 +40,52 @@ var _ datastore.Datastore = (*Datastore)(nil)
-		path:   path,
-		getDir: getDir,
-		sync:   sync,
+func New(path string, fun0 string, sync bool) (*Datastore, error) {
```
Okay, but I will still need the "auto" test case as that is used by the conversion utility.

kevina replied:
I didn't do this yet.  Do you basically want:
```
func New(path string, sync bool) (*Datastore, error) {
  NewWithFunc(path, "auto", sync)
}

func NewWithFunc(path string, fun0 string, sync bool) {...}
```

Is it that you don't like the special "auto" value?  I don't want to eliminate the use of that string as it makes writing the conversion function/utility easier, so using two separate functions won't really simplify any code.

kevina commented on flatfs.go:
```diff
@@ -38,37 +40,52 @@ var _ datastore.Datastore = (*Datastore)(nil)
-		str := padding + noslash
-		return str[len(str)-suffixLen:]
+	fun := ""
+	if fun0 == "auto" && fun1 == "auto" {
```
Okay.  I don't particular see the difference, but don't object either.  In the future I will try to remember that switch is generally preferred over multiple `else if`s.

Kubuxu commented:
What is the content of the SHARDING file directly? It isn't clear from the code?

kevina commented:
> What is the content of the SHARDING file directly? It isn't clear from the code?

Example file:
```
v1/next-to-last/2
```
As discussed in https://github.com/ipfs/go-ipfs/issues/3463

Kubuxu commented:
Could you change it to previously proposed: `/repo/flatfs/shard/v1/next-to-last/2`. There are no costs in it but it might make someone's day some time in future.

kevina commented:
@whyrusleeping agree with @Kubuxu? I have to problem doing it.

whyrusleeping commented:
Yeap, agreed with @Kubuxu

kevina commented:
> Could you change it to previously proposed: /repo/flatfs/shard/v1/next-to-last/2

Done.  Also refactored the parsing of the shard identifier string to make it more obvious what it going on.

whyrusleeping commented on flatfs.go:
```diff
@@ -38,37 +40,53 @@ var _ datastore.Datastore = (*Datastore)(nil)
+		return nil, fmt.Errorf("shard function not specified")
+	case fun0 == "auto":
+		fun = fun1
+	case fun1 == "auto":
```
I think we should separate the task of opening an existing datastore and creating a new one. That way we can remove a lot of the weirdness here

kevina replied:
I would rather not.  The "weirdness" is also there to allow opening an existing repo without the SHATDING file.

whyrusleeping replied:
mmm.... We're going to have to run a migration anyways. I think thats acceptable.

kevina replied:
@whyrusleeping, okay to do things the way you want I will basically have to redo this pull request and the conversion code will require some adjusting.  Let me know.

kevina replied:
I started work on this.  Will have it ready some time Wed.

kevina commented:
@whyrusleeping okay please have another look, there are two fixme, but I think this is close to what you want.  There is a separate `Create` and `Open` method.

This fix also makes the flatfs a little more robust, espacally after the fixme to check if an existing directory without a SHARDING file is not empty.

kevina commented on shard.go:
```diff
@@ -0,0 +1,138 @@
+	if err != nil {
+		return err
+	}
+	if str == IPFS_DEF_SHARD {
```
@whyrusleeping I don't think returning an error is a good idea.  Especially if the flatfs datastore could be used outside of IPFS.  I also don't like the idea of a warning for the same reason, but can add one if you still think it is a good idea.

I will separate the step from the writing of the sharding function.

kevina commented on flatfs_test.go:
```diff
@@ -214,40 +218,42 @@ func testStorage(p *params, t *testing.T) {
 	}
+	if fs.ShardFunc() == flatfs.IPFS_DEF_SHARD && !haveREADME {
+		t.Error("expected _README file")
+	} else if fs.ShardFunc() == flatfs.IPFS_DEF_SHARD && !haveREADME {
```
Oops.  Yes.  And I have a bug.  Will fix.

kevina commented on shard.go:
```diff
@@ -0,0 +1,138 @@
+	}
+	switch len(parts) {
+	case 3:
+		return shardId{version: parts[0], funName: parts[1], param: parts[2]}
```
Okay, I will see what I can do.  I originally had parseShardFunc return an error but decided to push all the error checking into Func().  Will try refactoring again and see if I can make it work better the way I think you want.

kevina replied:
Fixed now.

kevina commented:
@whyrusleeping I should have addressed all your concerns in the latest review except for returning an error if no README can be created because I am not sure that it was a good idea.  See my comment.

kevina commented:
Note the two special files are named:
```
  _README
  SHARDING
```
I use '_README' because that is what @jbenet wanted.  For consistency I can also rename 'SHARDING' to '_SHARDING' if you want.  Let me know.  For the record, I think using the special name '_README' is a good idea to help make it clear there is something special about it.

Kubuxu commented on flatfs.go:
```diff
@@ -38,37 +40,86 @@ var _ datastore.Datastore = (*Datastore)(nil)
-		getDir: getDir,
-		sync:   sync,
+var DatastoreExists = errors.New("Datastore already exist")
+var DatastoreDoesNotExist = errors.New("Datastore directory does not exist")
```
Please use one var block and `Err` name prefix.

kevina replied:
Okay.

Kubuxu commented on shard.go:
```diff
@@ -0,0 +1,133 @@
+		str = str[len(PREFIX):]
+	}
+	parts := strings.Split(str, "/")
+	if len(parts) == 3 {
```
Is the v1 optional? it doesn't have to be. Let's make it simpler.

kevina replied:
I don't agree (with "it doesn't have to be") but @whyrusleeping would probably say the same thing so I will change it.

Kubuxu commented on shard.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+func ReadShardFunc(dir string) (*ShardIdV1, error) {
+	buf, err := ioutil.ReadFile(filepath.Join(dir, "SHARDING"))
```
Let's make the name of the file constant as it is reused.

kevina replied:
Do we have a policy on this?  In this particular case it seams like an overkill to me.  We also don't seam to be very consistent about it elsewhere.

whyrusleeping replied:
We try to use constants for these sorts of things when we can, There are definitely cases where we forget to do so, but in general its a good idea

kevina replied:
Fine, Done.  However, I think it makes code harder to read and I also am horrible at naming the things.

Kubuxu commented on flatfs.go:
```diff
@@ -38,37 +40,86 @@ var _ datastore.Datastore = (*Datastore)(nil)
+		return nil, err
+	}
+
+	if err != nil {
```
Double err handler.

kevina replied:
Oops.  Will fix.

Kubuxu commented on shard.go:
```diff
@@ -0,0 +1,133 @@
+			return nil, fmt.Errorf("invalid prefix in shard identifier: %s", str)
+		}
+		str = str[len(PREFIX):]
+	}
```
Let's use TrimPrefix, and check if output string is different than the input, if it is then prefix was trimmed.

kevina replied:
Okay.

Kubuxu commented on shard.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+func Prefix(prefixLen int) ShardFunc {
+	padding := strings.Repeat("_", prefixLen)
```
It there a better way to do it? It seems clever but it might have performance penalty (string concat), and it is done for every key, potentially multiple time.

Kubuxu replied:
This applies to other ShardFuncs too.

kevina replied:
This has not changed.  The function was simply moved and copied.  Yeah there might be a better way, but I don't want to address it in this pull request.

kevina replied:
There might be yes.  But this code has not changed, just moved.  The same thing was also done before any of my shading changes.  So for now (at least this p.r.) I would rather leave it alone.

kevina replied:
There might be yes.  But this code has not changed, just moved.  The same thing was also done before any of my shading changes.  So for now (at least this p.r.) I would rather leave it alone.

[Sorry if this is a duplicate.  GitHub is being annoying.]

kevina commented:
@Kubuxu @whyrusleeping I should have addressed all your concerns.  I also implemented the fixme.

The only outstanding issue is returning an error if no README can be created because I am not sure that it was a good idea.  From my comment above:

> > Lets separate the step of writing the readme from writing the sharding function, and also make sure to error out or print a log message if we try to write a readme for an unsupported sharding type

> @whyrusleeping I don't think returning an error is a good idea. Especially if the flatfs datastore could be used outside of IPFS. I also don't like the idea of a warning for the same reason, but can add one if you still think it is a good idea.

Kubuxu commented on shard.go:
```diff
@@ -0,0 +1,135 @@
+
+func WriteReadme(dir string, id *ShardIdV1) error {
+	if id.String() == IPFS_DEF_SHARD {
+		err := ioutil.WriteFile(filepath.Join(dir, README_FN), []byte(IPFS_DEF_SHARD), 0444)
```
It should be `README_DEF_SHARD`

kevina replied:
Okay.  Told you I was bad a naming things. :)

kevina replied:
Oops.  Misunderstood and didn't see my own error.  Will fix.

Kubuxu commented on shard.go:
```diff
@@ -0,0 +1,135 @@
+	if err != nil {
+		return err
+	}
+	return nil
```
All above can be just `return err`.

Kubuxu commented on flatfs.go:
```diff
@@ -38,37 +40,90 @@ var _ datastore.Datastore = (*Datastore)(nil)
+		if err != nil {
+			return err
+		}
+		return nil
```
can be just `return err` too (no need for if).

Kubuxu commented on flatfs.go:
```diff
@@ -38,37 +40,90 @@ var _ datastore.Datastore = (*Datastore)(nil)
+
+	err := os.Mkdir(path, 0777)
+	if err != nil && !os.IsExist(err) {
+		return nil
```
Shouldn't it be `return err`?

kevina replied:
Yes.

kevina commented:
@Kubuxu anything else :)

kevina commented:
Note: Just rebased to clean up the commits, nothing else has changed.

kevina commented:
Note: I just pushed some more commits to avoid the use of strings wherever possible.
